pkg/trace/sampler/exception_sampler.go

```diff
@@ -0,0 +1,204 @@
+			e.report()
+		}
+	}()
+}
+
+// Stop stops reporting stats
+func (e *ExceptionSampler) Stop() {
+	e.tickStats.Stop()
+}
+
+func (e *ExceptionSampler) handlePriorityTrace(now time.Time, env string, t pb.Trace) {
+	expire := now.Add(priorityTTL)
+	for _, s := range t {
+		if !traceutil.HasTopLevel(s) && !traceutil.IsMeasured(s) {
+			continue
+		}
+		e.addSpan(expire, env, s)
+	}
+}
+
+func (e *ExceptionSampler) handleTrace(now time.Time, env string, t pb.Trace) bool {
+	var sampled bool
+	expire := now.Add(defaultTTL)
+	for _, s := range t {
+		if !traceutil.HasTopLevel(s) && !traceutil.IsMeasured(s) {
+			continue
+		}
+		if !sampled {
+			sampled = e.sampleSpan(now, env, s)
+			continue
+		}
+		e.addSpan(expire, env, s)
+	}
+	return sampled
+}
+
+func (e *ExceptionSampler) addSpan(expire time.Time, env string, s *pb.Span) {
+	shardSig := ServiceSignature{env, s.Service}.Hash()
+	b := e.loadSpanSeenSet(shardSig)
+	b.add(expire, s)
+}
+
+func (e *ExceptionSampler) sampleSpan(now time.Time, env string, s *pb.Span) bool {
+	var sampled bool
+	shardSig := ServiceSignature{env, s.Service}.Hash()
+	b := e.loadSpanSeenSet(shardSig)
+	sig := b.sign(s)
+	expire, ok := b.getExpire(sig)
+	if now.After(expire) || !ok {
+		sampled = e.limiter.Allow()
+		if sampled {
+			b.add(now.Add(defaultTTL), s)
+			atomic.AddInt64(&e.hits, 1)
+			traceutil.SetMetric(s, exceptionKey, 1)
+		} else {
+			atomic.AddInt64(&e.misses, 1)
+		}
+	}
+	return sampled
+}
+
+func (e *ExceptionSampler) loadSpanSeenSet(shardSig Signature) *seenSpans {
+	e.mu.RLock()
+	s, ok := e.seen[shardSig]
+	e.mu.RUnlock()
+	if ok {
+		return s
+	}
+	s = &seenSpans{expires: make(map[spanHash]time.Time)}
+	e.mu.Lock()
+	e.seen[shardSig] = s
+	e.mu.Unlock()
+	return s
+}
+
+func (e *ExceptionSampler) report() {
+	metrics.Count("datadog.trace_agent.sampler.exception.hits", atomic.SwapInt64(&e.hits, 0), nil, 1)
+	metrics.Count("datadog.trace_agent.sampler.exception.misses", atomic.SwapInt64(&e.misses, 0), nil, 1)
+}
+
+// seenSpans keeps record of a set of spans.
+type seenSpans struct {
+	mu sync.RWMutex
+	// expires contains expire time of each span seen.
+	expires map[spanHash]time.Time
+	// shrunk caracterize seenSpans when it's limited in size by capacityLimit.
+	shrunk bool
+}
+
+func (ss *seenSpans) add(expire time.Time, s *pb.Span) {
+	sig := ss.sign(s)
+	storedExpire, ok := ss.getExpire(sig)
+	if ok && expire.Sub(storedExpire) < ttlRenewalPeriod {
+		return
+	}
+	// slow path
+	ss.mu.Lock()
+	ss.expires[sig] = expire
+
+	// if cardinality limit reached, shrink
+	size := len(ss.expires)
+	if size > cardinalityLimit {
+		ss.shrink()
+	}
+	ss.mu.Unlock()
+}
+
+// shrink limits the cardinality of signatures considered and the memory usage.
+// This ensure that a service with high cardinality of resources does not consume
+// all sampling tokens. The cardinality limit matches a backend limit.
+// This function is not thread safe and should be called between locks
+func (ss *seenSpans) shrink() {
```

Comment:
Do we have tests for this function? Or if it's part of a larger functional test, do we know it's shrinking properly?

Reply:
Yes it's the TestCardinalityLimit test